.github/workflows/build-types.yml

```diff
@@ -0,0 +1,84 @@
+name: Build types
+
+on:
+  push:
+    branches:
+      - dev
+    paths:
+      - '.github/workflows/build-types.yml'
+
+      # We don't need to hunt for every possible case, these should cover almost all cases
+      - 'ui/package.json'
+      - 'ui/build/build.types.js'
+      - 'ui/build/build.javascript.js'
+
+      # Manually authored .d.ts files
+      - 'ui/types/**'
+      # JSON API files
+      - 'ui/src/**/*.json'
+      - '!ui/src/api-file-example.json'
+  pull_request:
+    types:
+      - opened
+      - synchronize
+    paths:
+      - '.github/workflows/build-types.yml'
+
+      # We don't need to hunt for every possible case, these should cover almost all cases
+      - 'ui/package.json'
+      - 'ui/build/build.types.js'
+      - 'ui/build/build.javascript.js'
+
+      # Manually authored .d.ts files
+      - 'ui/types/**'
+      # JSON API files
+      - 'ui/src/**/*.json'
+      - '!ui/src/api-file-example.json'
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  check:
+    runs-on: ubuntu-latest
+
+    permissions:
+      contents: read # to fetch code (actions/checkout)
+      actions: write # to upload artifacts (actions/upload-artifact)
+
+    defaults:
+      run:
+        working-directory: ui
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 16
+          # No caching because we are not checking lock files in
+
+      - name: Install dependencies
+        run: yarn
+
+      - name: Ensure types are formatted correctly
+        run: yarn format:types --write=false --check
+
+      - name: Build the types
+        run: yarn build js types
```

Comment:
This workflow runs on `push` and `pull_request` events.

The `push` event runs on our own commits, or when we merge a PR. So, unless we merge a malicious PR or do malicious commits ourselves, we are safe.

The `pull_request` event runs in the context of the PR branch, so it runs on the fork or wherever the PR is created from. That means if there is malicious code, they will affect their own repo. It doesn't have write access to our repo, so we can't create PR comments in the same workflow for example. Because of that limitation, we take the PR number and the build results and upload them as an artifact so we can process them in another workflow(_`./comment-build-results.yml`_) as commented below that file. So, basically, the worse a malicious user could do is to trick the build results, which would be noticeable if the user modified the build script suspiciously.
The problematic thing would be if we used `pull_request_target` and also did an explicit checkout of the user's code. In that scenario, we would check out the user's untrusted code, then give them access to our secrets and such. So, unless you do both at the same time, you are pretty much safe.

I'd suggest giving https://securitylab.github.com/research/github-actions-preventing-pwn-requests a read, which I put in the related workflow.

Reply:
Yep, I read that but still missed the part explaining that `pull_request` executes into the fork environment instead of ours, thanks for the detailed explanation :)